src/runtime/runtime.ts

```diff
@@ -647,33 +861,44 @@ class JoinNode implements Node {
   }
 
   exec(index:Index, input:Change, prefix:ID[], transaction:number, round:number, results:ID[][] = createArray()):boolean {
-    let ok = this.applyInput(input, prefix);
-    let countOfSolved = 0;
-    for(let elem of prefix) {
-      if(elem !== undefined) countOfSolved++;
-    }
-    let remainingToSolve = this.registerLength - countOfSolved;
-    let valid = this.presolveCheck(index, input, prefix, transaction, round);
-    if(!valid) {
-      return false;
-    } else if(!remainingToSolve) {
-      // if it is valid and there's nothing left to solve, then we've found
-      // a full result and we chould just continue
-      results.push(prefix.slice());
-      return true;
-    }
-
-    if(ok === ApplyInputState.fail) {
-      return false;
-    } else if(ok === ApplyInputState.pass && remainingToSolve) {
-      // For each node, find the new results that match the prefix.
-      this.genericJoin(index, prefix, transaction, round, results, remainingToSolve);
-      return true;
-    } else {
-      // If there is no affected prefix then tautologically there is no affected result, so we skip execution.
-      return true;
+    let didSomething = false;
+    this.inputState.constraintIx = 0;
+    // we need to apply input for each constraint individually and then run genericJoin to
+    // see if it produces any output. This handles the case where a single input might apply
+    // to several constraints in the block and we need to make sure we cover every permutation.
+    // @TODO: handle the case where all the constraints have applied a change.
+    this.applyInput(input, prefix);
+    let ok = this.inputState.state;
+    let ix = this.inputState.constraintIx;
+    let foo = 0;
+    while(ix < this.constraints.length) {
```

Comment:
This needs to be re-re-worked to happen over the set of valid input combinations.